src/Database/LSMTree/Internal/RawPage.hs

```diff
@@ -153,26 +165,74 @@ rawPageLookup !page !key
       | otherwise
       = LookupEntryNotPresent
 
+-- |
+-- __Time:__ \( \mathcal{O}\left( \log_2 n \right) \)
+-- where \( n \) is the number of entries in the 'RawPage'.
+--
+-- Return the least entry number in the 'RawPage' (if it exists)
+-- which is greater than or equal to the supplied 'SerialisedKey'.
+--
+-- The following law always holds \( \forall \mathtt{key} \mathtt{page} \):
+-- >>> maybe True (key <=) (getRawPageIndexKey . rawPageIndex page . id   =<< rawPageFindKey page key)
+-- >>> maybe True (key > ) (getRawPageIndexKey . rawPageIndex page . pred =<< rawPageFindKey page key)
+rawPageFindKey
+    :: RawPage
+    -> SerialisedKey
+    -> Maybe Word16  -- ^ entry number of first entry greater or equal to the key
+rawPageFindKey !page !key
+  | dirNumKeys == 1 = lookup1
+  | otherwise       = case bisectPageToKey page key of
+    KeyNotInPage   -> Nothing
+    KeyFoundAt _ i -> Just $ fromIntegral i
+
+  where
+    !dirNumKeys = rawPageNumKeys page
+
+    lookup1
+      | key <= rawPageKeyAt page 0 = Just 0
+      | otherwise                  = Nothing
+
+-- | This data-type facilitates the code reuse of 'bisectPageToKey' between the
+-- and 'rawPageLookup' and 'rawPageFindKey'.
+--
+-- NOTE: Be sure to enable UnboxedSums to improve efficiency of this data-type.
+data BinarySearchResult
+  = KeyNotInPage
+  | KeyFoundAt {-# UNPACK #-} !Ordering  {-# UNPACK #-} !Int
```

Comment:
The warnings here are correct, you cannot unpack the `Ordering`, but on the other hand given that it's an enum, there's little advantage either.

Note also that we only pattern match on EQ or not EQ, so we don't need the full `Ordering`. A bool would be enough.